components/quickbooks/actions/download-pdf/download-pdf.js

```diff
@@ -0,0 +1,64 @@
+const quickbooks = require("../../quickbooks.app");
+const fs = require("fs");
+const { promisify } = require("util");
+const stream = require("stream");
+
+module.exports = {
+  name: "Download PDF",
+  description: "Download an invoice, bill, purchase order or other QuickBooks entity as a PDF and save it in [Pipedream's temporary file system](https://pipedream.com/docs/workflows/steps/code/nodejs/working-with-files/#the-tmp-directory) for use in a later step.",
+  key: "quickbooks-download-pdf",
+  version: "0.0.1",
+  type: "action",
+  props: {
+    quickbooks,
+    entity: {
+      type: "string",
+      label: "Document Type",
+      description: null,
+      options: [
+        "CreditMemo",
+        "Estimate",
+        "Invoice",
+        "Payment",
+        "PurchaseOrder",
+        "RefundReceipt",
+        "SalesReceipt",
+      ],
+    },
+    id: {
+      type: "string",
+      label: "Record ID",
+      description: null,
```

Comment:
To help users find an entity's **Record ID**, I'd recommend adding a [description](https://pipedream.com/docs/components/guidelines/#descriptions) that explains where to find the ID or links to relevant QuickBooks API docs.

Alternatively/additionally, use [async options](https://pipedream.com/docs/components/guidelines/#async-options) (with label/value definitions) so users can make selections from a drop down menu.
For example:
- `id` prop in `download-pdf.js`
```js
    id: {
      type: "string",
      label: "Record ID",
      description: null,
      async options({ page }) {
        const limit = 100;
        const { entity } = this;
        if (!entity) {
          return [];
        }
        let response;
        try {
          response = await this.quickbooks.queryEntity(this, entity, {
            startPosition: 1 + (page) * limit,
            maxResults: limit,
          });
        } catch (err) {
          return [];
        }
        if (!response[entity]) {
          return [];
        }
        return response[entity].map((e) => ({
          label: this.quickbooks.getOptionLabel(e, entity),
          value: e.Id,
        }));
      },
    },
```
- New `methods` in `quickbooks.app.js`
```js
    getOptionLabel(entity, entityName) {
      switch (entityName) {
      case "CreditMemo":
        return entity.DocNumber
          ? `# ${entity.DocNumber}`
          : entity.Id;
      case "Estimate":
        return `${entity.totalAmount} (# ${entity.DocNumber})`;
      case "Invoice":
        return `Invoice # ${entity.DocNumber}`;
      case "Payment":
        return `${entity.CustomerRef?.name ?? ""} (${entity.Id})`;
      case "PurchaseOrder":
        return `${entity.VendorRef?.name ?? ""} (# ${entity.DocNumber})`;
      case "RefundReceipt":
        return entity.DocNumber
          ? `# ${entity.DocNumber}`
          : entity.Id;
      case "SalesReceipt":
        return entity.DocNumber
          ? `# ${entity.DocNumber}`
          : entity.Id;
      default:
        return entity.Id;
      }
    },
    async queryEntity($, entityName, config = {}) {
      const {
        startPosition = 1,
        maxResults = 100,
      } = config;
      if (!WEBHOOK_ENTITIES.includes(entityName)) {
        throw new Error(`"${entityName}" is not a valid QuickBooks entity`);
      }
      // Query operations docs:
      // https://developer.intuit.com/app/developer/qbo/docs/learn/explore-the-quickbooks-online-api/data-queries
      const query = `
        SELECT * FROM ${entityName}
        ORDER BY Metadata.LastUpdatedTime DESC
        STARTPOSITION ${startPosition} MAXRESULTS ${maxResults}`;
      const companyId = this.companyId();
      return (await this._makeRequest($, {
        path: `company/${companyId}/query`,
        params: {
          query,
        },
      })).QueryResponse;
    },
```